osquery/filesystem/posix/xattrs.cpp

```diff
@@ -70,73 +130,92 @@ Status getExtendedAttributeNameList(std::vector<std::string>& name_list,
-    name_list.push_back(std::move(name));
+    name_list.emplace_back(std::move(name));
   }
 
-  return Status::success();
+  return name_list;
 }
 
-Status getExtendedAttributeValue(ExtendedAttributeValue& value,
-                                 const std::string& path,
-                                 const std::string& name) {
-  value = {};
-
-  auto buffer_size = getxattr(path.c_str(), name.c_str(), nullptr, 0);
+XAttrValueResult getExtendedAttributeValue(int fd, const std::string& name) {
+  auto buffer_size = fgetxattr(fd, name.c_str(), nullptr, 0);
   if (buffer_size == 0) {
-    return Status::success();
-
+    return XAttrValueResult::success({});
   } else if (buffer_size < 0) {
-    return Status::failure(
-        "Failed to determine the length of the extended attribute named '" +
-        name + "' from the following path: " + path);
+    return createError(XAttrValueError::NoLength);
   }
 
-  ExtendedAttributeValue buffer(static_cast<std::size_t>(buffer_size), 0U);
+  ExtendedAttributeValue value(static_cast<std::size_t>(buffer_size), 0U);
 
-  auto err = getxattr(path.c_str(), name.c_str(), buffer.data(), buffer.size());
+  auto err = fgetxattr(fd, name.c_str(), value.data(), value.size());
   if (err == 0) {
-    return Status::success();
+    return value;
 
   } else if (err < 0) {
-    return Status::failure("Failed to acquire the extended attribute named '" +
-                           name + "' from the following path: " + path);
+    return createError(XAttrValueError::Get);
   }
 
   if (err < buffer_size) {
-    buffer.resize(static_cast<std::size_t>(err));
+    value.resize(static_cast<std::size_t>(err));
 
   } else if (err > buffer_size) {
-    return Status::failure("Failed to acquire the extended attribute named '" +
-                           name + "' from the following path: " + path +
-                           ". The extended attribute value has been changed "
-                           "while it was being acquired.");
+    return createError(XAttrValueError::SizeChanged);
   }
 
-  value = std::move(buffer);
-  return Status::success();
+  return value;
 }
 
-Status getExtendedAttributes(ExtendedAttributeMap& xattr_map,
-                             const std::string& path) {
-  xattr_map = {};
+XAttrGetResult getExtendedAttributes(const std::string& path) {
+  int fd = open(path.c_str(), O_RDONLY);
 
-  std::vector<std::string> name_list;
-  auto status = getExtendedAttributeNameList(name_list, path);
-  if (!status.ok()) {
-    return status;
+  if (fd < 0) {
+    if (errno == ENOENT) {
+      return createError(XAttrGetError::NoFile);
+    }
+
+    std::string error_message =
+        "Failed to open file to read extended attributes at the "
+        "following path: ";
+    error_message += path;
+
+    return XAttrGetResult::failure(XAttrGetError::GenericError,
+                                   std::move(error_message));
   }
 
+  auto fd_guard = scope_guard::create([&] { close(fd); });
```

Comment:
Nice! I didn't realize we had this

Reply:
Yeah, it's a bit of a catch all solution. Ideally we would be able to use PlatformFile for this, but as it's currently written it adds overhead/does operations one doesn't always want.